crates/biome_service/src/workspace/server.rs

```diff
@@ -491,84 +497,41 @@ impl WorkspaceServer {
             })
     }
 
-    /// Extracts embedded content from HTML elements containing foreign
-    /// languages.
-    ///
-    /// This function walks the HTML syntax tree to find all `<script>` and
-    /// `<style>` elements and extracts their content for offset-aware
-    /// parsing.
+    fn get_language_snippets(&self, path: &Utf8Path) -> Vec<EmbeddedSnippets> {
+        let documents = self.documents.pin();
+        documents
+            .get(path)
+            .map(|document| document.embedded_snippets.clone())
+            .unwrap_or_default()
+    }
+
+    /// Parses the language snippets if the current language implements the capability `parser.parse_embedded_nodes`
     fn parse_embedded_language_snippets(
         &self,
+        project_key: ProjectKey,
         path: &BiomePath,
-        parse: &AnyParse,
-        settings: &Settings,
-        file_source_index: usize,
-        source_index_fn: impl Fn(DocumentFileSource) -> usize,
-    ) -> Result<EmbeddedLanguageSnippets, WorkspaceError> {
-        let (source, html_root) = match self.get_source(file_source_index) {
-            Some(source @ DocumentFileSource::Html(_)) => match HtmlRoot::cast(parse.syntax()) {
-                Some(html_root) => (source, html_root),
-                None => return Ok(Default::default()),
-            },
-            _ => return Ok(Default::default()),
+        source: &DocumentFileSource,
+        root: &AnyParse,
+        cache: &mut NodeCache,
+    ) -> Result<Vec<EmbeddedSnippets>, WorkspaceError> {
+        let mut embedded_nodes = Vec::new();
+        let capabilities = self.get_file_capabilities(path);
+        let Some(parse_embedded) = capabilities.parser.parse_embedded_nodes else {
+            return Ok(Default::default());
         };
+        let settings = self
+            .projects
+            .get_settings_based_on_path(project_key, path)
+            .ok_or_else(WorkspaceError::no_project)?;
```

Comment:
It is better to revert this so that `settings` is passed as an argument again instead of queried here, so that we don't do redundant lookups for the settings within a single workspace method. I did recently create a PR for this, so it's possible you had a merge conflict on this.